pyproject.toml

```
@@ -1,10 +1,225 @@
 [build-system]
 requires = [
-    "setuptools >= 35.0.2",
-    "wheel >= 0.29.0",
+    "hatchling",
+    "hatch-fancy-pypi-readme >= 22.5.0",
     "incremental >= 21.3.0",
 ]
-build-backend = "setuptools.build_meta"
+build-backend = "hatchling.build"
+
+[project]
+name = "Twisted"
+dynamic = [
+    "version",
+    "readme",
+]
+description = "An asynchronous networking framework written in Python"
+license = "MIT"
+# When updating this value, make sure our CI matrix includes a matching minimum version.
+requires-python = ">=3.7.1"
+authors = [
+    { name = "Twisted Matrix Laboratories", email = "twisted-python@twistedmatrix.com" },
+]
+classifiers = [
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3 :: Only",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+]
+dependencies = [
+    "zope.interface >= 4.4.2",
+    "constantly >= 15.1",
+    "incremental >= 21.3.0",
+    "Automat >= 0.8.0",
+    "hyperlink >= 17.1.1",
+    "attrs >= 19.2.0",
+    "typing_extensions >= 3.6.5",
+    "twisted-iocpsupport >= 1.0.2, <2; platform_system == 'Windows'",
+]
+
+[project.optional-dependencies]
+# This is list of minimum dependencies required to run the tests.
+# The plan is to remove the `test` extra requirement and skip tests that
+# depend on them
+# To prevent trunk failures we should pin all dev dependencies to minor
+# versions.
+# We trust semantic versioning and auto-upgrading to a bugfix release
+# should be OK.
+test = [
+    "cython-test-exception-raiser >= 1.0.2, <2",
+    "PyHamcrest >= 1.9.0",
+    "hypothesis ~= 6.0",
+]
+
+# List of dependencies required to build the documentation and test the
+# release scripts and process.
+dev-release = [
```

The spelling of the extras - specifically, with `_` and not `-` - is part of an explicitly public interface for the package.  Folks all over are expected and required to write things like `pip install twisted[macos-platform]` or put `twisted[macos-platform]` into their package metadata, their CI configurations, etc.

Do we _have_ to change the names?  If not, it's worth considering just not doing it.  If we do, how do we provide a transitional period in line with the project's stated compatibility policies?

---

As mentioned by JP, we need to keep the same extra names.

```suggestion
dev_release = [
```

or if we introduce new names and plan to deprecated them, then add backward compatiblity with a warning...but I think it's too much work.

So I hope that the new standard allows both `dev_release` and `dev-release`
and for Twisted we will continue to use underscore.

For now, this patch introducing `dev-release` without support for the old `dev_release` is a backward incompatible change and I think it should be considered a blocker for the merge.

I hope this can use `dev_release` and then we should be able to land this.

For `pip` end user, there will be no change.

---

> I hope that the new standard allows both `dev_release` and `dev-release`

It doesn't, so this is fortuitous timing since Hatchling is just the first to implement PEP 685 which says to always write normalized names, and this would have actually been an issue soon since the `setuptools` version was unbounded here.

> then add backward compatiblity with a warning

Prior art: `urllib3` did [this](https://twitter.com/sethmlarson/status/1558076150206926850) but this strategy would be out of my control

> For `pip` end user, there will be no change

We could wait until `pip` itself normalizes which is coming soon, but then, how long?

---

> We could wait until pip itself normalizes which is coming soon, but then, how long?

We should continue shipping unnormalized, version 2.2 metadata for at least a year, per our compatability policy. This will also save us a lot of grief as Pips that normalize trickle out.

---

If  the "modern" tools don't support using things like `dev_release` this looks like to me like the "modern" tools don't care about backward compatibility.

And we should not encourage the usage of backward incompatible tools.

I am not -1 for "modern" tooling, but it looks to me like is designed only for new proejcts.

I guess there is no hurry to use the "modern" tools, and we can wait until the old tools are no longer supported or until the modern tools are backward compatible.

---

Fixed, I had to implement https://hatch.pypa.io/latest/config/metadata/#allowing-ambiguous-features

---

Thank you for the extra feature @ofek ! Nice to drive the ecosystem forward :).  Given that it's addressed, I'll resolve this part of the conversation.